client/extensions/woocommerce/app/settings/email/mailchimp/setup-mailchimp.js

```diff
@@ -0,0 +1,315 @@
+/**
+ * External dependencies
+ */
+import React from 'react';
+import { connect } from 'react-redux';
+import { pick, some, isEmpty } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import { localize } from 'i18n-calypso';
+import Dialog from 'components/dialog';
+import RequiredPluginsInstallView from 'woocommerce/app/dashboard/required-plugins-install-view';
+import ProgressIndicator from 'components/wizard/progress-indicator';
+import ProgressBar from 'components/progress-bar';
+import {
+	submitMailChimpApiKey,
+	submitMailchimpStoreInfo,
+	submitMailchimpCampaignDefaults,
+	submitMailchimpNewsletterSettings
+} from 'woocommerce/state/sites/settings/email/actions.js';
+import { isSubbmittingApiKey, isApiKeyCorrect } from 'woocommerce/state/sites/settings/email/selectors';
+import LogIntoMailchimp from './setup-steps/log-into-mailchimp.js';
+import StoreInfoStep from './setup-steps/store-info.js';
+import CampaignDefaultsStep from './setup-steps/campaign-defaults.js';
+import NewsletterSettings from './setup-steps/newsletter-settings.js';
+import KeyInputStep from './setup-steps/key-input.js';
+import { getStoreLocation } from 'woocommerce/state/sites/settings/general/selectors';
+import { getCurrencyWithEdits } from 'woocommerce/state/ui/payments/currency/selectors';
+import { getCurrentUserEmail } from 'state/current-user/selectors';
+import { getSiteTimezoneValue } from 'state/selectors';
+
+const LOG_INTO_MAILCHIMP_STEP = 'log_into';
+const KEY_INPUT_STEP = 'key_input';
+const STORE_INFO_STEP = 'store_info';
+const CAMPAIGN_DEFAULTS_STEP = 'campaign_defaults';
+const NEWSLETTER_SETTINGS_STEP = 'newsletter_settings';
+const STORE_SYNC = 'sync';
+
+const steps = {
+	[ LOG_INTO_MAILCHIMP_STEP ]: { number: 0, nextStep: KEY_INPUT_STEP },
+	[ KEY_INPUT_STEP ]: { number: 1, nextStep: STORE_INFO_STEP },
+	[ STORE_INFO_STEP ]: { number: 2, nextStep: CAMPAIGN_DEFAULTS_STEP },
+	[ CAMPAIGN_DEFAULTS_STEP ]: { number: 3, nextStep: NEWSLETTER_SETTINGS_STEP },
+	[ NEWSLETTER_SETTINGS_STEP ]: { number: 4, nextStep: STORE_SYNC }
+};
+
+const storSettingsRequriredFields = [ 'store_name', 'store_street', 'store_city', 'store_state',
+	'store_postal_code', 'store_country', 'store_phone', 'store_locale', 'store_timezone',
+	'store_currency_code', 'admin_email' ];
+
+const campaignDefaultsRequiredFields = [ 'campaign_from_name', 'campaign_from_email', 'campaign_subject',
+		'campaign_language', 'campaign_permission_reminder' ];
+
+class MailChimpSetup extends React.Component {
+
+	constructor( props ) {
+		super( props );
+		// make this react to the real phase the execution is.
+		this.state = {
+			step: LOG_INTO_MAILCHIMP_STEP,
+			settings: this.prepareDefaultValues( this.props.settings ),
+			settings_values_missing: false,
+			api_key_input: this.props.settings.mailchimp_api_key || '',
+		};
+	}
+
+	componentWillReceiveProps( nextProps ) {
+		if ( nextProps.settings.mailchimp_lists && ! this.state.settings.mailchimp_lists ) {
```

Comment:
Another null check on `nextProps.settings`? Or do we set a default prop of settings to `{}` to ensure all is good? There are more spots further down in this method that we should either check for the existence of settings or ensure we have a default empty object just to be safe.

Reply:
^ same